src/Thentos/Config.hs

```diff
@@ -1,312 +1,199 @@
+  :*> Maybe ("default_user" :> DefaultUserConfig' :>: "A user that is created if the user table is empty.")
+
+defaultThentosConfig :: ToConfig ThentosConfigUntagged Maybe
+defaultThentosConfig =
+      Just Run
+  :*> NothingO
+  :*> NothingO
+  :*> NothingO
+  :*> Just defaultSmtpConfig
+  :*> NothingO
+
+type HttpConfig = Tagged (NoDesc (ToConfigCode HttpConfig'))
+type HttpConfig' =
+      Maybe ("bind_schema"   :> HttpSchema)
+  :*>       ("bind_host"     :> ST)
+  :*>       ("bind_port"     :> Int)
+  :*> Maybe ("expose_schema" :> HttpSchema)
+  :*> Maybe ("expose_host"   :> ST)
+  :*> Maybe ("expose_port"   :> Int)
+
+type HttpProxyConfig = Tagged (NoDesc (ToConfigCode HttpProxyConfig'))
+type HttpProxyConfig' =
+            ("service_id" :> ST)
+  :*>       ("http"       :> HttpConfig')
+  :*> Maybe ("url_prefix" :> ST)
+
+type SmtpConfig = Tagged (NoDesc (ToConfigCode SmtpConfig'))
+type SmtpConfig' =
+      Maybe ("sender_name"    :> ST)  -- FIXME: use more specific type 'Network.Mail.Mime.Address'
+  :*>       ("sender_address" :> ST)
+  :*>       ("sendmail_path"  :> ST)
+  :*>       ("sendmail_args"  :> [ST])
+
+defaultSmtpConfig :: ToConfig (NoDesc (ToConfigCode SmtpConfig')) Maybe
+defaultSmtpConfig =
+      NothingO
+  :*> Nothing
+  :*> Just "/usr/sbin/sendmail"
+  :*> Just ["-t"]
+
+type DefaultUserConfig = Tagged (NoDesc (ToConfigCode DefaultUserConfig'))
+type DefaultUserConfig' =
+            ("name"     :> ST)  -- FIXME: use more specific type?
+  :*>       ("password" :> ST)  -- FIXME: use more specific type?
+  :*>       ("email"    :> ST)  -- FIXME: use more specific type 'Network.Mail.Mime.Address'
+  :*> Maybe ("roles"    :> [Role])
+
+
+-- * leaf types
+
+data Command = Run | RunA3 | ShowDB
+  deriving (Eq, Ord, Show, Enum, Bounded, Typeable, Generic)
+
+instance Aeson.ToJSON Command where toJSON = Aeson.gtoJson
+instance Aeson.FromJSON Command where parseJSON = Aeson.gparseJson
+
+data HttpSchema = Http | Https
+  deriving (Eq, Ord, Enum, Bounded, Typeable, Generic)
+
+instance Show HttpSchema where
+    show Http = "http"
+    show Https = "https"
+
+instance Aeson.ToJSON HttpSchema where toJSON = Aeson.gtoJson
+instance Aeson.FromJSON HttpSchema where parseJSON = Aeson.gparseJson
+
+
+-- * driver
+
+printConfigUsage :: IO ()
+printConfigUsage = do
+    -- ST.putStrLn $ docs (Proxy :: Proxy ThentosConfigDesc)
+    ST.putStrLn $ docs (Proxy :: Proxy (ToConfigCode ThentosConfig'))
+
+
+getConfig :: FilePath -> IO ThentosConfig
+getConfig configFile = do
+    sources <- sequence
+        [ ConfigFileYaml <$> SBS.readFile configFile
+        , ShellEnv       <$> getEnvironment
+        , CommandLine    <$> getArgs
+        ]
+    logger DEBUG $ "config sources:\n" ++ ppShow sources
+
+    case configify' (TaggedM defaultThentosConfig) sources :: Result ThentosConfigUntagged of
+        Left e -> do
+            logger CRITICAL $ "error parsing config: " ++ show e
```

Comment:
Yes, I am now, but you just had it right in front of you, and it took me 5 minutes to check out the branch build it again, and get the error myself.  I don't know the answer to this error immediately, so I had to go through that anyway, but you didn't know that.

Anyway, my bad.  I submitted a branch for review appearently without having started the main executable, and with a sketchy test suite....

I'll look into it.